packages/core/src/runtime-feature.ts

```diff
@@ -0,0 +1,170 @@
+import { deferred, IDeferredPromise } from 'promise-assist';
+
+// interface RuntimeFeature<F, E> {
+//     [RUN](engine: RuntimeEngine): Promise<void>;
+//     [DISPOSE](engine: RuntimeEngine): Promise<void>;
+//     addRunHandler(fn: () => unknown, envName: string): void;
+//     addOnDisposeHandler(fn: DisposeFunction, envName: string): void;
+//     api: Running<F, E>;
+//     dependencies: RunningFeatures<F['dependencies'], E>;
+// }
+
+/**
+ * Represents a currently running feature instance.
+ **/
+export class RuntimeFeature<T extends FeatureDescriptor, ENV extends AnyEnvironment> {
+    private running = false;
+    private runHandlers = new SetMultiMap<string, () => unknown>();
+    private disposeHandlers = new SetMultiMap<string, DisposeFunction>();
+    private disposing: IDeferredPromise<void> | undefined;
+
+    constructor(
+        public feature: T,
+        public api: Running<T, ENV>,
+        public dependencies: RunningFeaturesV2<T['dependencies'], ENV>
+    ) {}
+
+    public addRunHandler(fn: () => unknown, envName: string) {
+        this.runHandlers.add(envName, fn);
+    }
+    public addOnDisposeHandler(fn: DisposeFunction, envName: string) {
+        this.disposeHandlers.add(envName, fn);
+    }
+    public async [RUN](engine: RuntimeEngine): Promise<void> {
+        if (this.running) {
+            return;
+        }
+        this.running = true;
+        const runPromises: Array<unknown> = [];
+        for (const envName of engine.referencedEnvs) {
+            for (const dep of this.feature.dependencies) {
+                runPromises.push(engine.runFeature(dep));
+            }
+            const envRunHandlers = this.runHandlers.get(envName);
+            if (envRunHandlers) {
+                for (const handler of envRunHandlers) {
+                    runPromises.push(handler());
+                }
+            }
+        }
+
+        await Promise.all(runPromises);
+    }
+    public async [DISPOSE](engine: RuntimeEngine) {
+        const {
+            entryEnvironment: { env: envName },
+        } = engine;
+        if (this.disposing) {
+            return this.disposing.promise;
+        }
+        this.disposing = deferred();
+        // THIS IS WRONG!
+        for (const dep of this.feature.dependencies) {
+            await engine.disposeFeature(dep);
+        }
+        const featureDisposeHandlers = this.disposeHandlers.get(envName) || new Set();
+        for (const handler of featureDisposeHandlers) {
+            await handler();
+        }
+        this.running = false;
+        return this.disposing.resolve();
+    }
+}
+
+export function createFeatureRuntime<F extends FeatureDescriptor, E extends AnyEnvironment>(
+    feature: F,
+    runningEngine: RuntimeEngine<E>
+): RuntimeFeature<F, E> {
+    const {
+        features,
+        runOptions,
+        referencedEnvs,
+        entryEnvironment: { env: envName },
+    } = runningEngine;
+
+    const deps: any = {};
```

Comment:
types to this and the next 5 lines

Reply:
This is harder then it looks...not the be able to write a type there but later to avoid the `as`.
I will look at this in the next iteration on engine.